src/app/ReadHandler.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines read handler for a CHIP Interaction Data model
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/MessageDef/EventPath.h>
+#include <app/ReadHandler.h>
+
+namespace chip {
+namespace app {
+CHIP_ERROR ReadHandler::Init(Messaging::ExchangeManager * apExchangeMgr, InteractionModelDelegate * apDelegate)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    // Error if already initialized.
+    VerifyOrExit(mpExchangeMgr == nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    mpExchangeMgr     = apExchangeMgr;
+    mpExchangeCtx     = nullptr;
+    mpDelegate        = apDelegate;
+    mSuppressResponse = true;
+    mGetToAllEvents   = true;
+    mState            = HandlerState::Initialized;
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadHandler::Reset()
+{
+    ClearExistingExchangeContext();
+    MoveToState(HandlerState::Initialized);
+}
+
+void ReadHandler::Shutdown()
+{
+    ClearExistingExchangeContext();
+    MoveToState(HandlerState::Uninitialized);
+    mpExchangeMgr = nullptr;
+    mpDelegate    = nullptr;
+}
+
+CHIP_ERROR ReadHandler::ClearExistingExchangeContext()
+{
+    if (mpExchangeCtx != nullptr)
+    {
+        mpExchangeCtx->Abort();
+        mpExchangeCtx = nullptr;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void ReadHandler::OnMessageReceived(Messaging::ExchangeContext * apExchangeContext, const PacketHeader & aPacketHeader,
+                                    const PayloadHeader & aPayloadHeader, System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferHandle response;
+
+    mpExchangeCtx = apExchangeContext;
+
+    err = ProcessReadRequest(std::move(aPayload));
```

Comment:
Is the assumption here that caller checked this is actually a Read Request message?  If so, please document.

Reply:
yes, just documented